user/src/lib.rs

```diff
@@ -0,0 +1,318 @@
+    }
+
+    /// Pops open a browser window for the update URL. This is less encountered by users as time goes
+    /// by, but still used.
+    pub fn update_app(&self) -> bool {
+        if let Err(error) = open::that_detached("https://slippi.gg/downloads?update=true") {
+            tracing::error!(?error, "Failed to open update URL");
+            return false;
+        }
+
+        true
+    }
+
+    /// Returns whether we have an authenticated user - i.e, whether we were able
+    /// to find/load/parse their `user.json` file.
+    pub fn is_logged_in(&self) -> bool {
+        self.get(|user| user.uid != "")
+    }
+
+    /// During matchmaking, we may opt to force-overwrite the latest version to
+    /// account for errors that can happen when the user tries to update.
+    pub fn overwrite_latest_version(&self, version: String) {
+        self.set(|user| {
+            user.latest_version = version;
+        });
+    }
+
+    /// Logs the current user out and removes their `user.json` from the filesystem.
+    pub fn logout(&mut self) {
+        self.set(|user| *user = UserInfo::default());
+
+        if let Err(error) = std::fs::remove_file(self.user_json_path.as_path()) {
+            tracing::error!(?error, "Failed to remove user.json on logout");
+        }
+
+        let mut watcher = self.watcher.lock().expect("Unable to acquire watcher lock on user logout");
+
+        watcher.logout();
+    }
+}
+
+/// Checks for the existence of a `user.json` file and, if found, attempts to load and parse it.
+///
+/// This returns a `bool` value so that the background thread can know whether to stop checking.
+fn attempt_login(http_client: &Agent, user: &Arc<Mutex<UserInfo>>, user_json_path: &PathBuf, slippi_semver: &str) -> bool {
+    match std::fs::read_to_string(user_json_path) {
+        Ok(contents) => match serde_json::from_str::<UserInfo>(&contents) {
+            Ok(mut info) => {
+                info.sanitize();
+
+                let uid = info.uid.clone();
+                {
+                    let mut lock = user.lock().expect("Unable to lock user in attempt_login");
+
+                    *lock = info;
+                }
+
+                overwrite_from_server(http_client, user, uid, slippi_semver);
+                return true;
+            },
+
+            Err(error) => {
+                tracing::error!(?error, "Unable to parse user.json");
+                return false;
+            },
+        },
+
+        Err(error) => {
+            // A not-found file just means they haven't logged in yet... presumably.
+            if error.kind() != std::io::ErrorKind::NotFound {
+                tracing::error!(?error, "Unable to read user.json");
+            }
+
+            return false;
+        },
+    }
+}
+
+/// The core payload that represents user information. This type is expected to conform
+/// to the same definition that the remote server uses.
+#[derive(Debug, Default, serde::Deserialize)]
+pub struct APIResponse {
```

Comment:
does this need to be public? it doesn't seem like it, but if it does then i'd like a better name so you can tell at a glance what you're getting.
maybe something like `GetUserInfoResp`

Reply:
Ah, it does _not_ need to be public, no - good catch.

I think this stems from originally having the API call in a separate module while I was mocking it out. Easy swap tho.

Reply:
Done (renamed for clarity and removed `pub`).